dask/array/utils.py

```diff
@@ -388,6 +388,35 @@ def zeros_like_safe(a, shape, **kwargs):
         return np.zeros(shape, **kwargs)
 
 
+def _array_like_safe(np_func, da_func, a, like, **kwargs):
+    if like is a and hasattr(a, "__array_function__"):
+        return a
+
+    if isinstance(like, Array):
+        return da_func(a, **kwargs)
+    elif isinstance(a, Array):
+        a = a.compute(scheduler="sync")
```

Comment:
Why do we compute in this case? Is there an example of how this should work?

Reply:
Documented in https://github.com/dask/dask/pull/7162/files#diff-65211e64fa680da306e9612b92c60f557365507d46486325f0e7e04359bce64fR410-R413 . Would you like me to duplicate the comment here?

Reply:
Thanks for the pointer. No that shouldn't be needed. Was just reading the code too quickly

Where does `a` come from in this case? Am trying to understand if there may be any unexpected side-affects by computing here

Reply:
`a` is an array passed to `array_safe`, e.g., `dask.array.utils.array_safe(a)` (analogous to `np.array(a)`).

Reply:
Would it be possible to call this on `meta` instead?

Reply:
No, this is the actual value that gets passed down to NumPy, who later dispatches to a downstream library. Think of the following case:

```python
In [1]: import dask.array as da, cupy as cp, numpy as np

In [2]: rs = da.random.RandomState(RandomState=cp.random.RandomState)

In [3]: a = rs.random((5,))

In [4]: a
Out[4]: dask.array<random_sample, shape=(5,), dtype=float64, chunksize=(5,), chunktype=cupy.ndarray>

In [5]: np.array(a, like=cp.array(()))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-0af7c34305ab> in <module>
----> 1 np.array(a, like=cp.array(()))

cupy/core/core.pyx in cupy.core.core.ndarray.__array_function__()

/datasets/pentschev/miniconda3/envs/rn-110-0.18.210203/lib/python3.8/site-packages/cupy/_creation/from_data.py in array(obj, dtype, copy, order, subok, ndmin)
     39
     40     """
---> 41     return core.array(obj, dtype, copy, order, subok, ndmin)
     42
     43

cupy/core/core.pyx in cupy.core.core.array()

cupy/core/core.pyx in cupy.core.core.array()

cupy/core/core.pyx in cupy.core.core._send_object_to_gpu()

/datasets/pentschev/src/dask/dask/array/core.py in __array__(self, dtype, **kwargs)
   1448             x = x.astype(dtype)
   1449         if not isinstance(x, np.ndarray):
-> 1450             x = np.array(x)
   1451         return x
   1452

cupy/core/core.pyx in cupy.core.core.ndarray.__array__()

TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```

In the case above, because `a` is a Dask array backed by CuPy, whenever it's passed down to CuPy a call to `__array__` will be attempted, which isn't allowed by CuPy. However, I realized that's probably overkill as this won't be a limitation for all downstream libraries. Therefore I limited that for now to CuPy arrays.

Reply:
Ok that makes sense. Thanks for clarifying! 😄